### ISSUE TITLE
**OOM** in OpCode `POW`

### ISSUE THREAD

**Issue by vang1ong7ang** [neo-project/neo#2536]
```
1201fe00a34a9d9e12001ba3a3
```

source code: <https://github.com/lazynode/Tanya/pull/10>

Essentially becasue `BigInteger.Pow(value, exponent)` mallocs a lot of mem. See:

https://github.com/neo-project/neo-vm/blob/5da8f71782812d9aa298c819c429e314f37938ab/src/neo-vm/ExecutionEngine.cs#L846-L852

**Comment by vang1ong7ang** [comment]
@roman-khimov it could be a general problem and there might be similar risks in neo-go.

**Comment by roman-khimov** [comment]
Yeah, it doesn't seem to OOM, but it allocates a substantial chunk and already works on calculating that for about 3 minutes.

**Comment by roman-khimov** [comment]
I think we can make input checking a bit more complex here:
 * allow any positive `exponent` for `value` in `[-1, 0, 1]`
 * only allow `exponent < MaxShift` for any other `value`

We can of course just implement the second, but who knows these smart contracts, maybe there is some use case for the first case.

**Comment by vang1ong7ang** [comment]
> * allow any positive `exponent` for `value` in `[-1, 0, 1]`
> * only allow `exponent < MaxShift` for any other `value`

i think it's exactly best solution

**Comment by dusmart** [comment]
> * allow any positive `exponent` for `value` in `[-1, 0, 1]`
> * only allow `exponent < MaxShift` for any other `value`

Great idea. C# stdlib do the same thing in `pow`. 

https://referencesource.microsoft.com/#System.Numerics/System/Numerics/BigInteger.cs

**Comment by vang1ong7ang** [comment]
BTW, IMO **MODPOW** which introduced by @igormcoelho here <https://github.com/neo-project/neo-vm/issues/77> could be more useful than **POW**.
